internal/encryption/keyprovider/gcp_kms/config.go

```diff
@@ -0,0 +1,138 @@
+func ReadPathOrContents(poc string) (string, error) {
+	if len(poc) == 0 {
+		return poc, nil
+	}
+
+	path := poc
+	if path[0] == '~' {
+		var err error
+		path, err = homedir.Expand(path)
+		if err != nil {
+			return path, err
+		}
+	}
+
+	if _, err := os.Stat(path); err == nil {
+		contents, err := os.ReadFile(path)
+		if err != nil {
+			return string(contents), err
+		}
+		return string(contents), nil
+	}
+
+	return poc, nil
+}
+
+func (c Config) Build() (keyprovider.KeyProvider, keyprovider.KeyMeta, error) {
+	// This mirrors the gcp remote state backend
+
+	// Apply env defaults if nessesary
+	c.Credentials = stringAttrEnvFallback(c.Credentials, "GOOGLE_CREDENTIALS")
+	c.AccessToken = stringAttrEnvFallback(c.AccessToken, "GOOGLE_OAUTH_ACCESS_TOKEN")
+	c.ImpersonateServiceAccount = stringAttrEnvFallback(c.ImpersonateServiceAccount, "GOOGLE_BACKEND_IMPERSONATE_SERVICE_ACCOUNT")
+	c.ImpersonateServiceAccount = stringAttrEnvFallback(c.ImpersonateServiceAccount, "GOOGLE_IMPERSONATE_SERVICE_ACCOUNT")
+	c.KMSKeyName = stringAttrEnvFallback(c.KMSKeyName, "GOOGLE_KMS_ENCRYPTION_KEY")
+
+	ctx := context.Background()
+
+	var opts []option.ClientOption
+	var credOptions []option.ClientOption
+
+	if c.AccessToken != "" {
+		tokenSource := oauth2.StaticTokenSource(&oauth2.Token{
+			AccessToken: c.AccessToken,
+		})
+		credOptions = append(credOptions, option.WithTokenSource(tokenSource))
+	} else if c.Credentials != "" {
+		// to mirror how the provider works, we accept the file path or the contents
+		contents, err := ReadPathOrContents(c.Credentials)
+		if err != nil {
+			return nil, nil, &keyprovider.ErrInvalidConfiguration{Message: "Error loading credentials", Cause: err}
+		}
+
+		if !json.Valid([]byte(contents)) {
+			return nil, nil, &keyprovider.ErrInvalidConfiguration{Message: "the string provided in credentials is neither valid json nor a valid file path"}
+		}
+
+		credOptions = append(credOptions, option.WithCredentialsJSON([]byte(contents)))
+	}
+
+	// Service Account Impersonation
+	if c.ImpersonateServiceAccount != "" {
+		ts, err := impersonate.CredentialsTokenSource(ctx, impersonate.CredentialsConfig{
+			TargetPrincipal: c.ImpersonateServiceAccount,
+			Scopes:          []string{"https://www.googleapis.com/auth/cloudkms"}, // I can't find a smaller scope than this...
+			Delegates:       c.ImpersonateServiceAccountDelegates,
+		}, credOptions...)
+
+		if err != nil {
+			return nil, nil, &keyprovider.ErrInvalidConfiguration{Cause: err}
+		}
+
+		opts = append(opts, option.WithTokenSource(ts))
+
+	} else {
+		opts = append(opts, credOptions...)
+	}
+
+	opts = append(opts, option.WithUserAgent(httpclient.OpenTofuUserAgent(version.Version)))
+
+	svc, err := kms.NewKeyManagementClient(ctx, opts...)
+	if err != nil {
+		return nil, nil, &keyprovider.ErrInvalidConfiguration{Cause: err}
+	}
+
+	if c.KeySize < 1 {
+		return nil, nil, &keyprovider.ErrInvalidConfiguration{Message: "key_size must be at least 1"}
+	}
+	if c.KeySize > 1024 {
```

Comment:
I think we should validate the key size similarly as in pbkdf2. I believe it only allows 16, 24, and 32, but please double-check.